### PR TITLE
Bump gioui version

### PR DESCRIPTION
**Pull request by ukane-philemon**
closes https://github.com/crypto-power/cryptopower/issues/173

This PR bumps the gioui version to version 0.7 and adds fixes for notable breaking changes.

- Buttons now accept a `gtx layout.Context`: To pass `gtx` to action that need them, I have refactored and add gtx to `pg.HandleUserInteractions(gtx C)` and move `modal.Handle` (modals action handler)  to their `.Layout(gtx)` method. This conforms to the existing pattern we already follow in the codebase. 

- Most of the events and UI components in v0.7 have changed, the way events are handled has also changed

 - This is not an exhaustive list of changes see https://gioui.org/news for more info on breaking changes in various "major" gioui versions.

- Also some bugs have been fixed in this PR
 
 Review notice:

- This is a major change to the entire UI logic and event handling of the entire app. It needs to be test very carefully.
- I have checked the application again, but please test the application and double check the following parts:
+ UI when running on IOS and Android
+ Editor related events such as copy/paste, change, submit event...
+ Click out event and hide keyboard on mobile devices

### CODE REVIEW

**Comment by dreacot** [comment]
this pr needs to come alive again

**Comment by ukane-philemon** [comment]
Whoops! That's a hell of a conflict lol. I'll get to this soon.

**Comment by JustinBeBoy** [comment]
I am finishing it

**Comment by ukane-philemon** [comment]
> I am finishing it

Thanks!

File: main.go [hunk 1]
```diff
@@ -40,6 +40,15 @@ func main() {
...
 
+	// f, _ := os.Create("cpu.pprof")
+	// pprof.StartCPUProfile(f)
+
+	// mem, _ := os.Create("mem.pprof")
+	// runtime.GC()
+	// if err := pprof.WriteHeapProfile(mem); err != nil {
+	// 	fmt.Println("could not write memory profile: ", err)
+	// }
+
```
Comment (dmigwi) [inline main.go:51]: This looks like some test code you need to delete.

File: main.go [hunk 2]
```diff
@@ -110,6 +119,8 @@ func main() {
 	go func() {
 		// Wait until we receive the shutdown request.
 		<-win.Quit
+		// pprof.StopCPUProfile()
+		// mem.Close()
```
Comment (dmigwi) [inline main.go:123]: Same to this. Delete it, if it is no longer needed

File: ui/cryptomaterial/clickable_list.go [hunk 3]
```diff
@@ -54,14 +54,14 @@ func (cl *ClickableList) handleClickables(count int) {
 	}
 
 	for index, clickable := range cl.clickables {
-		for clickable.Clicked() {
+		for clickable.Clicked(gtx) {
```
Comment (dmigwi) [inline ui/cryptomaterial/clickable_list.go:57]:
```suggestion
		if clickable.Clicked(gtx) {
```

File: ui/cryptomaterial/collapsible.go [hunk 4]
```diff
@@ -85,7 +85,7 @@ func (t *Theme) CollapsibleWithOption() *CollapsibleWithOption {
 }
 
 func (c *Collapsible) Layout(gtx C, header, body func(C) D) D {
-	for c.button.Clicked() {
+	for c.button.Clicked(gtx) {
```
Comment (dmigwi) [inline ui/cryptomaterial/collapsible.go:88]:
```suggestion
	if c.button.Clicked(gtx) {
```

File: ui/cryptomaterial/dropdown.go [hunk 5]
```diff
@@ -173,12 +173,12 @@ func (d *DropDown) Len() int {
...
 }
 
-func (d *DropDown) handleEvents() {
+func (d *DropDown) handleEvents(gtx C) {
 	if d.expanded {
 		for i := range d.items {
 			index := i
 			item := d.items[index]
-			for item.clickable.Clicked() {
+			for item.clickable.Clicked(gtx) {
```
Comment (dmigwi) [inline ui/cryptomaterial/dropdown.go:181]:
```suggestion
			if item.clickable.Clicked(gtx) {
```

File: ui/cryptomaterial/dropdown.go [hunk 6]
```diff
@@ -187,18 +187,18 @@ func (d *DropDown) handleEvents() {
 			}
 		}
 	} else {
-		for d.clickable.Clicked() {
+		for d.clickable.Clicked(gtx) {
```
Comment (dmigwi) [inline ui/cryptomaterial/dropdown.go:190]:
```suggestion
		if d.clickable.Clicked(gtx) {
```

File: ui/cryptomaterial/dropdown.go [hunk 7]
```diff
@@ -187,18 +187,18 @@ func (d *DropDown) handleEvents() {
...
 }
 
-func (d *DropDown) Changed() bool {
+func (d *DropDown) Changed(gtx C) bool {
 	if d.expanded {
 		for i := range d.items {
 			index := i
 			item := d.items[index]
-			for item.clickable.Clicked() {
+			for item.clickable.Clicked(gtx) {
```
Comment (dmigwi) [inline ui/cryptomaterial/dropdown.go:201]:
```suggestion
			if item.clickable.Clicked(gtx) {
```

File: ui/cryptomaterial/editor.go [hunk 8]
```diff
@@ -65,15 +72,21 @@ type Editor struct {
 	m2 unit.Dp
 	m5 unit.Dp
 
-	clickable *widget.Clickable
+	// clickable *widget.Clickable
```
Comment (dmigwi) [inline ui/cryptomaterial/editor.go:75]: Remove this comment

File: ui/cryptomaterial/hoverable.go [hunk 9]
```diff
@@ -22,39 +21,36 @@ func (h *Hoverable) Hovered() bool {
...
-	for _, e := range gtx.Events(h) {
-		ev, ok := e.(pointer.Event)
+	start := h.hovered
+	for {
+		ev, ok := gtx.Event(pointer.Filter{
+			Target: h,
+			Kinds:  pointer.Enter | pointer.Leave,
+		})
 		if !ok {
-			continue
```
Comment (dmigwi) [inline ui/cryptomaterial/hoverable.go:33]: Nicee!! This might have messed up someone!

File: ui/cryptomaterial/modal.go [hunk 10]
```diff
@@ -31,19 +31,21 @@ type Modal struct {
...
+	isDisabled           bool
+	showScrollBar        bool
+	isMobileView         bool
+	isFirstDisplay       bool
+	firstLoadWithContext func(gtx C)
 }
 
-func (t *Theme) ModalFloatTitle(id string, isMobileView bool) *Modal {
-	mod := t.Modal(id, isMobileView)
+func (t *Theme) ModalFloatTitle(id string, isMobileView bool, firstLoad ...func(gtx C)) *Modal {
```
Comment (dmigwi) [inline ui/cryptomaterial/modal.go:42]: You don't need a variadic parameter here because it requires always checking if it was passed. Just pass a function with an empty body if you don't want to execute anything.

kindly also add a comment on why the extra parameter is necessary.
```suggestion

func (t *Theme) ModalFloatTitle(id string, isMobileView bool, firstLoad func(gtx C)) *Modal {
```

**Review by dmigwi — CHANGES_REQUESTED** [review]
Here is my partial review. Please check it. I will proceed with the review later on!

File: ui/cryptomaterial/modal.go (hunk 10 above)
Comment (JustinBeBoy) [reply ui/cryptomaterial/modal.go:44]: Your approach is good too, but it is a option parameter, it is reason it should is  variadic parameter, it should be handled in one place instead of passing functions with empty bodies pointlessly if not needed

File: ui/page/components/wallet_account_selector.go [hunk 11]
```diff
@@ -522,34 +522,32 @@ func (sm *selectorModal) accountValidator(accountIsValid func(*sharedW.Account)
...
-				switch item := selectorItem.item.(type) {
-				case *sharedW.Account:
-					if sm.onAccountClicked != nil {
-						sm.onAccountClicked(item)
-					}
-				case sharedW.Asset:
-					if sm.onWalletClicked != nil {
-						sm.onWalletClicked(item)
-					}
+func (sm *selectorModal) Handle(gtx C) {
```
Comment (dmigwi) [inline ui/page/components/wallet_account_selector.go:525]: check if source has been assigned.

File: ui/page/components/wallet_setup_page.go [hunk 12]
```diff
@@ -114,11 +114,9 @@ func NewCreateWallet(l *load.Load, walletCreationSuccessCallback func(), assetTy
 
 	pg.passwordEditor = l.Theme.EditorPassword(new(widget.Editor), values.String(values.StrSpendingPassword))
 	pg.passwordEditor.Editor.SingleLine, pg.passwordEditor.Editor.Submit = true, true
-	pg.passwordEditor.Hint = values.String(values.StrSpendingPassword)
 
-	pg.confirmPasswordEditor = l.Theme.EditorPassword(new(widget.Editor), values.String(values.StrSpendingPassword))
+	pg.confirmPasswordEditor = l.Theme.EditorPassword(new(widget.Editor), values.String(values.StrConfirmSpendingPassword))
 	pg.confirmPasswordEditor.Editor.SingleLine, pg.confirmPasswordEditor.Editor.Submit = true, true
-	pg.confirmPasswordEditor.Hint = values.String(values.StrConfirmSpendingPassword)
```
Comment (dmigwi) [inline ui/page/components/wallet_setup_page.go:121]: Why delete the hint?

File: ui/page/governance/consensus_page.go [hunk 13]
```diff
@@ -245,8 +247,9 @@ func (pg *ConsensusPage) HandleUserInteractions() {
 										layout.Flexed(0.9, pg.Theme.Body1(host).Layout),
 										layout.Flexed(0.1, func(gtx C) D {
 											return layout.E.Layout(gtx, func(gtx C) D {
-												if pg.copyRedirectURL.Clicked() {
-													clipboard.WriteOp{Text: host}.Add(gtx.Ops)
+												if pg.copyRedirectURL.Clicked(gtx) {
+													// clipboard.WriteOp{Text: host}.Add(gtx.Ops)
+													gtx.Execute(clipboard.WriteCmd{Data: io.NopCloser(strings.NewReader(host))})
```
Comment (dmigwi) [inline ui/page/governance/consensus_page.go:252]: Delete this comment.

File: ui/page/governance/proposal_details_page.go [hunk 14]
```diff
@@ -245,8 +247,9 @@ func (pg *ProposalDetails) HandleUserInteractions() {
 										layout.Flexed(0.1, func(gtx C) D {
 											return layout.E.Layout(gtx, func(gtx C) D {
 												return layout.Inset{Top: values.MarginPadding7}.Layout(gtx, func(gtx C) D {
-													if pg.copyRedirectURL.Clicked() {
-														clipboard.WriteOp{Text: host}.Add(gtx.Ops)
+													if pg.copyRedirectURL.Clicked(gtx) {
+														// clipboard.WriteOp{Text: host}.Add(gtx.Ops)
```
Comment (dmigwi) [inline ui/page/governance/proposal_details_page.go:251]: delete comment

**Review by dmigwi — CHANGES_REQUESTED** [review]

File: ui/cryptomaterial/modal.go (hunk 10 above)
Comment (JoeGruffins) [reply ui/cryptomaterial/modal.go:44]: You can also just pass nil I think if there's only one function and you don't need it.

File: ui/page/components/asset_selector.go [hunk 15]
```diff
@@ -224,17 +224,15 @@ func (ats *AssetTypeSelector) buildExchangeItems() []*AssetTypeItem {
...
 
-func (atm *assetTypeModal) Handle() {
-	if atm.eventQueue != nil {
-		for _, assetTypeItem := range atm.assetTypeItems {
-			for assetTypeItem.clickable.Clicked() {
-				atm.onAssetTypeClicked(assetTypeItem)
-				atm.Dismiss()
-			}
+func (atm *assetTypeModal) Handle(gtx C) {
+	for _, assetTypeItem := range atm.assetTypeItems {
```
[the remark this answers is not in the document]
Comment (JustinBeBoy) [reply ui/page/components/asset_selector.go:228]: we don't need check nil for eventSource because it can't nil

File: ui/page/components/wallet_account_selector.go (hunk 11 above)
Comment (JustinBeBoy) [reply ui/page/components/wallet_account_selector.go:525]: we don't need check nil for eventSource because it can't nil

File: ui/page/components/wallet_setup_page.go (hunk 12 above)
Comment (JustinBeBoy) [reply ui/page/components/wallet_setup_page.go:121]: hit include when init Editor in l.Theme.EditorPassword()

File: ui/cryptomaterial/modal.go (hunk 10 above)
Comment (JustinBeBoy) [reply ui/cryptomaterial/modal.go]: oh, thank you @dmigwi @JoeGruffins

**Review by JoeGruffins — APPROVED** [review]

File: ui/window.go [hunk 16]
```diff
@@ -189,6 +190,19 @@ func (win *Window) HandleEvents() {
...
+	go func() {
+		for {
+			ev := win.Event()
+			events <- ev
+			<-acks
+			if _, ok := ev.(giouiApp.DestroyEvent); ok {
+				return
+			}
+		}
+	}()
```
[the remark this answers is not in the document]
Comment (JustinBeBoy) [reply ui/window.go:205]: On new gioui version win.Event() not return chan for event so we need to create chan for it

File: ui/page/dcrdex/dex_onboarding_page.go [hunk 17]
```diff
@@ -1142,7 +1144,7 @@ func (pg *DEXOnboarding) connectServerAndPrepareForBonding() {
 		xc, err = dexClient.GetDEXConfig(pg.bondServer.url, pg.bondServer.cert)
 	}
 	if err != nil {
-		pg.notifyError(fmt.Errorf("Error retrieving DEX server info: %w", err).Error())
+		pg.notifyError(fmt.Sprintf("Error retrieving DEX server info: %s", err.Error()))
```
Comment (dmigwi) [inline ui/page/dcrdex/dex_onboarding_page.go:1147]: Was this change necessary?

File: ui/page/dcrdex/market.go [hunk 18]
```diff
@@ -1606,34 +1561,93 @@ func (pg *DEXMarketPage) HandleUserInteractions() {
...
+
+	if pg.orderHistoryBtn.Clicked(gtx) {
+		pg.orders = nil // clear orders
+		pg.openOrdersDisplayed = false
+		go pg.refreshOrders()
+	}
+
+	if pg.seeFullOrderBookBtn.Clicked(gtx) {
+		// TODO: display full order book
+		log.Info("Display full order book")
```
Comment (dmigwi) [inline ui/page/dcrdex/market.go:1609]: log an error like this " button click listener for full order book view is not implemented"

File: ui/page/exchange/exchange_selector.go [hunk 19]
```diff
@@ -226,17 +226,17 @@ func (es *ExSelector) buildExchangeItems(server ...instantswap.Server) []*exchan
...
 
-func (em *exchangeModal) Handle() {
-	if em.eventQueue != nil {
-		for _, exchangeItem := range em.exchangeItems {
-			for exchangeItem.clickable.Clicked() {
-				em.onExchangeClicked(exchangeItem.item)
-				em.Dismiss()
-			}
+func (em *exchangeModal) Handle(gtx C) {
+	// if em.eventQueue != nil {
```
Comment (dmigwi) [inline ui/page/exchange/exchange_selector.go:230]: Remove this comment

File: ui/page/exchange/exchange_selector.go [hunk 20]
```diff
@@ -45,7 +45,7 @@ type exchangeModal struct {
 	onExchangeClicked func(*Exchange)
 	exchangeList      layout.List
 	exchangeItems     []*exchangeItem
-	eventQueue        event.Queue
+	eventSource       input.Source
```
Comment (dmigwi) [inline ui/page/exchange/exchange_selector.go:48]: remove this variable if its no longer in use.

File: ui/page/exchange/frequency_selector.go [hunk 21]
```diff
@@ -39,7 +39,7 @@ type frequencyModal struct {
 	onFrequencyClicked func(*frequencyItem)
 	frequencyList      layout.List
 	frequencyItems     []*frequencyItem
-	eventQueue         event.Queue
+	eventSource        input.Source
```
Comment (dmigwi) [inline ui/page/exchange/frequency_selector.go:42]: remove this variable if its no longer in use.

File: ui/page/governance/proposal_details_page.go [hunk 22]
```diff
@@ -207,21 +209,21 @@ func (pg *ProposalDetails) getProposalItemWidgets() *proposalItemWidgets {
...
 		if len(pg.assetWallets) == 0 {
 			pg.displayCreateWalletModal(libutils.DCRWalletAsset)
 			return
 		}
 		pg.ParentWindow().ShowModal(newVoteModal(pg.Load, pg.proposal))
 	}
 
-	for pg.viewInPoliteiaBtn.Clicked() {
+	if pg.viewInPoliteiaBtn.Clicked(gtx) {
 		host := "https://proposals.decred.org/record/" + pg.proposal.Token
```
Comment (dmigwi) [inline ui/page/governance/proposal_details_page.go:227]: Make this URLs, constants declared at the top of the page.

File: ui/page/governance/treasury_page.go [hunk 23]
```diff
@@ -153,7 +153,7 @@ func (pg *TreasuryPage) HandleUserInteractions() {
 		pg.ParentWindow().ShowModal(infoModal)
 	}
 
-	for pg.viewGovernanceKeys.Clicked() {
+	if pg.viewGovernanceKeys.Clicked(gtx) {
 		host := "https://github.com/decred/dcrd/blob/master/chaincfg/mainnetparams.go#L477"
 		if pg.AssetsManager.NetType() == libwallet.Testnet {
```
Comment (dmigwi) [inline ui/page/governance/treasury_page.go:158]: Make this URLs constants declared at the top of the page.

File: ui/page/send/page.go [hunk 24]
```diff
@@ -676,12 +677,12 @@ func (pg *Page) OnDismiss() {
 	pg.OnNavigatedFrom()
 }
 
-// KeysToHandle returns an expression that describes a set of key combinations
-// that this page wishes to capture. The HandleKeyPress() method will only be
+// KeysToHandle returns a Filter's slice that describes a set of key combinations
+// that this modal wishes to capture. The HandleKeyPress() method will only be
```
Comment (dmigwi) [inline ui/page/send/page.go:681]: Why the modal reference here?

File: ui/page/wallet/single_wallet_main_page.go [hunk 25]
```diff
@@ -352,39 +353,39 @@ func (swmp *SingleWalletMasterPage) HandleUserInteractions() {
...
+	if swmp.hideBalanceButton.Clicked(gtx) {
 		swmp.isBalanceHidden = !swmp.isBalanceHidden
 		swmp.selectedWallet.SetBoolConfigValueForKey(sharedW.HideBalanceConfigKey, swmp.isBalanceHidden)
 	}
 }
 
-// KeysToHandle returns an expression that describes a set of key combinations
-// that this page wishes to capture. The HandleKeyPress() method will only be
+// KeysToHandle returns a Filter's slice that describes a set of key combinations
+// that this modal wishes to capture. The HandleKeyPress() method will only be
```
Comment (dmigwi) [inline ui/page/wallet/single_wallet_main_page.go:370]: Why the modal reference?

File: ui/window.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (dmigwi) [reply ui/window.go:205]: I don't think this new implementation is necessary. Check the docs here: https://github.com/gioui/gio/blob/86349775b7734bf72fbda72106c3dfd32929e4ef/app/doc.go#L51

File: ui/window.go [hunk 26]
```diff
@@ -198,19 +213,18 @@ func (win *Window) HandleEvents() {
 		case <-win.IsShutdown:
 			// backend processes shutdown is complete, exit UI process too.
 			return
-		case e := <-win.Events():
+		case e := <-events:
```
Comment (dmigwi) [inline ui/window.go:216]: Move this implementation to default switch case where you will use a loop to recursively check for new events from `win.Event()` function until `DestroyEvent` is detected.

**Review by dmigwi — CHANGES_REQUESTED** [review]
Here is an extra set of reviews to consider.

File: ui/page/components/wallet_account_selector.go [hunk 27]
```diff
@@ -55,7 +55,7 @@ type selectorModal struct {
 	walletsList        layout.List
 	selectorItems      []*SelectorItem // A SelectorItem can either be a wallet or account
 	assetType          []utils.AssetType
-	eventQueue         event.Queue
+	eventSource        input.Source
```
[the remark this answers is not in the document]
Comment (JustinBeBoy) [reply ui/page/components/wallet_account_selector.go:58]: good catch

File: ui/page/dcrdex/dex_onboarding_page.go (hunk 17 above)
Comment (JustinBeBoy) [reply ui/page/dcrdex/dex_onboarding_page.go:1147]: because we don't return error, so why we need create error for it, additionally, it will cause a warning when the returned message is capitalized

File: ui/page/dcrdex/market.go (hunk 18 above)
Comment (JustinBeBoy) [reply ui/page/dcrdex/market.go:1609]: hm.. it's old code, not in this PR's scope, but I will fix it

File: ui/window.go (hunk 26 above)
Comment (JustinBeBoy) [reply ui/window.go:216]: why we need do that? please check this example on gioui: https://github.com/gioui/gio-example/blob/4aa16fb00aeeffc801f08c049eddc4087f69b322/kitchen/kitchen.go#L136

File: ui/window.go [hunk 28]
```diff
@@ -189,6 +190,20 @@ func (win *Window) HandleEvents() {
 		win.Quit <- struct{}{}
 	}
 
+	// Create window chan event and listen events from window event
+	events := make(chan event.Event)
+	acks := make(chan struct{})
```
Comment (dmigwi) [inline ui/window.go:195]:
```suggestion
	acks := make(chan struct{},2)
```

Making `acks` channel buffered and have capacity of 2, prevents the channel blocking that happens when `DestroyEvent` is detected. For all the other events you write once to `acks` channel before reading but when `DestroyEvent` is detected you attempt to write to the channel twice. The second write will block the app shutdown if a buffered channel is not used.

**Review by dmigwi — COMMENTED** [review]
Apart from the buffered channel issue, everything else seems to work just okay!

**Review by dmigwi — APPROVED** [review]
The code looks good and works fine on my Mac!

**Review by JustinBeBoy — APPROVED** [review]

**Review by dreacot — APPROVED** [review]
